app/services/create_work_package_service.rb

```diff
@@ -0,0 +1,52 @@
+#-- encoding: UTF-8
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class CreateWorkPackageService
+  attr_accessor :user, :project, :work_package
+
+  def initialize(user:, project:, send_notifications: true)
+    self.user = user
+    self.project = project
+
+    WorkPackageObserver.instance.send_notification = send_notifications
+  end
+
+  def create
+    hash = {
+      project: project,
+      author: user,
+      type: project.types.where(is_default: true).first || project.types.first
+    }
+    self.work_package = project.add_work_package(hash)
+  end
+
+  def save
+    work_package.save
```

Comment:
I thought about the semantics of this service. and now it feels weird.

Should you be able to create more than one work package using the *same* service?

This is currently supported in the following way:

````ruby
while i_want_more_work_packages do
  service.create
  change_attributes_to_my_delight(service.work_package) # I could also store the wp in a local
  service.save
end
````

That means the service keeps track of the created work package and calling save will save the last WP created.

This also means the following snippets won't work:

````ruby
service.create.subject = 'My new subject'
service.create.description = 'My cool description'
service.save
````

````ruby
1..10 do |i|
  wp = service.create
  wp.subject = "Subject #{i}"
end
service.save
````

At first I had a hard time to figure out what I dislike about your approach, so I tried to remember principles of clean code and could not find the name of what I disliked, but now I found it!

The service smells like it utilizes [temporal coupling](http://blog.ploeh.dk/2011/05/24/DesignSmellTemporalCoupling/): A caller needs to know that he first calls `create` and then he can call `save`, everything else will not work.

You can either avoid such a coupling entirely **or** make it explicit. In my opinion that leaves two ways out:

1. Make the service stateless and the coupling explicit:

````ruby
wp = service.create
service.save(wp) # you can't call save without having a work package
````

2. Keep the service stateful, but remove the coupling by allowing only one WP per service:

````ruby
service.work_package.subject = 'foo'
service.save # uses the lazily initialized WP, if the client does not call `work_package`, everything will have default values
````

@ulferts I've written quite some text here, but I think you are one of the people pushing the idea of "services" forward, so I'd like you to have a look and an opinion too...